### DISCUSSION TITLE
Is it possible to generate a DELETE ... USING with LEFT JOIN on PostgreSQL

### DISCUSSION THREAD

**Discussion by lelit** [sqlalchemy/sqlalchemy#10178] [Usage Questions]
Hi, I have to perform a `DELETE` on a quite big table, returning the deleted records so that I can archive them in an external storage.
For historical reasons, the table is actually split into two halves, with a "_one-to-zero-or-one_" relation between them, so this is what I came up:
```python
actionlog_2 = actionlog.alias('al2')
stmt = (sa.delete(action)
        .where(
            (sa.func.age(action.c.creation_date) > sa.text(f"'{days_to_keep} days'"))
            &
            ((actionlog.c.action_id == action.c.id)
             |
             (~sa.exists(
                 sa.select(actionlog_2.c.id)
                 .where(actionlog_2.c.action_id == action.c.id)))))
        .returning(action.c.creation_date,
                   action.c.name,
                   action.c.value,
                   actionlog.c.log))
```

which generates something like
```sql
DELETE FROM public.action
USING public.actionlog
WHERE age(public.action.creation_date) > '18250 days'
  AND (public.actionlog.action_id = public.action.id
       OR NOT (EXISTS (SELECT al2.id 
                       FROM public.actionlog AS al2 
                       WHERE al2.action_id = public.action.id)))
RETURNING public.action.creation_date,
          public.action.name,
          public.action.value,
          public.actionlog.log
```

Although that is working as expected, trying out with hand-written SQL I found that a statement like the following works as well and is sensibly faster
```sql
DELETE FROM public.action
USING public.action AS a LEFT JOIN public.actionlog AS al ON al.action_id = a.id
WHERE age(public.action.creation_date) > '18250 days'
  AND public.action.id = a.id
RETURNING public.action.creation_date,
          public.action.name,
          public.action.value,
          al.log
```

I tried several approaches, but I could not find a way inject that `LEFT JOIN` into the `delete()`, because it does not accept an explicit `using` keyword but AFAICT it composes the `USING` clause analyzing the `where()` conditions.

If this is doable (possibly with SA 1.4.x), can you point me in the right direction?

**Comment by zzzeek** [comment]
we dont have this, and the potential feature add is discussed somewhat vaguely at #8130

**Reply by lelit** [reply]
Thank you, I missed that issue.

**Reply by zzzeek** [reply]
ive clarified the subject to make it easier to find and also proposed a new method `using()` to keep it simple
